informational/req-res-reliability.md

```diff
@@ -0,0 +1,62 @@
+tags: [informational]
+editor: Oleksandr Kozlov <oleksandr@status.im>
+contributors:
+- Prem Chaitanya Prathi <prem@status.im>
+- Danish Arora <danish@status.im>
+---
+
+## Abstract
+This RFC describes set of instructions used across different [WAKU2](https://github.com/vacp2p/rfc-index/blob/7b443c1aab627894e3f22f5adfbb93f4c4eac4f6/waku/standards/core/10/waku2.md?plain=1#L3) implementations for improved reliability in request-response protocols such as [WAKU2-LIGHTPUSH](https://github.com/vacp2p/rfc-index/blob/7b443c1aab627894e3f22f5adfbb93f4c4eac4f6/waku/standards/core/19/lightpush.md?plain=1#L3C11-L3C26) and [WAKU2-FILTER](https://github.com/vacp2p/rfc-index/blob/7b443c1aab627894e3f22f5adfbb93f4c4eac4f6/waku/standards/core/12/filter.md?plain=1#L3).
+
+## Motivation
+
+Descriptions of mentioned protocols do not define some of the real world use cases that are oftenly observed in unreliable network environment. Such use cases can be: recovery from offline state, decrease rate of missed messages, increase probability of messages being broadcasted within the network.
+
+## Suggestions
+
+### Node health
+
+As a useful metric to define and implement for determining quality of provided service by a node:
+- unhealthy - no peer connections are available regardless of protocol;
+- minimally healthy:
+  - Relay has less than 4 peers connected;
+  - Filter and LightPush has one per each peer connection available;
+- sufficiently healthy:
+  - Relay has minimum 4 peers connected;
+  - more than 1 connection in Filter and at least 2 connections available in LightPush;
+
+### Peers and connection management
+
+- Each protocols should retain a pool of reliable peers. In case a protocol failed to use any peer more than once - connection to it should be dropped and new peer should be added to the pool instead. 
+
+- During discovery of new peers it is better to filter unwonted out based on ENR / multiaddress. For example in some cases `circuit-relay` addresses are not needed when we try to find and connect to peers directly.
+
+- When peer is discovered second time, we need to be sure to keep connection information up to date in Peer Store.
+
+### Light Push 
+
+- While sending message with Light Push - it is advised to use more than 1 peer in order to increase changes of delivering message.
+
+- If sending message is failed to all of the peers - node should try to re-send message after some interval and continue doing so while OK response is received. 
+
+### Filter
+
+- To decrease chances of missing messages a node can initiate more than one subscription through Filter protocol to the same content topic and filter out duplicates. This will increase bandwidth consumption and would depend on the information exchanged under content topic in use.
+
+- In case a node goes offline while having an active subscription - it is important to do ping again right after node appears online. In case ping fails - re-subscribe request should be fired to a new peer.
```

Comment:
Interesting approach, but i have seen that it is practically not possible because detecting a node went offline takes time ~20-30seconds in which time the peer ping might end up failing and causing subscriptions to get removed.  
Also, note that Filter-ping doesn't have enough information to know which are the subscriptions currently active with a peer..it only indicates that there is a subscription with it. 
Currently in status-go/go-waku the way it is implemented is as soon as offline status is identified all subscriptions are cleared off. When node comes back online new subscriptions are created. 

At the same time, i am curious to know if we have taken the apporach mentioned here in js-waku what was the efficacy with node connectivity and did we not run into any issues?

Reply:
> detecting a node went offline takes time ~20-30seconds

that is not the case for browser, we can identify that it is offline almost instantly 

> When node comes back online new subscriptions are created.

that is interesting and this is what can kinda happen, as the described approach in the worst case will re-initiate all the subscriptions

> did we not run into any issues

it is still experimental but on my local repros I observe the issue when subscription is dropped by a service node, that is happening somewhere after 10-15 mins.
Considering that we might just not do a ping if a node was offline for more than 10 minutes, but the problem here is it is not controllable behavior, meaning there is no CLI flag specified and it is not described in Filter RFC so it is totally up to an implementation.

Reply:
to that, I would define how to use store to recover.
Probably something like that (terminology can be improved):

- timeout: the underlying protocol timeout. 30s for TCP, 1s/0s for websocket ("instant"). This caps how fast one can detect a disconnection
- interval: time between filter pings
- t0: successful ping
- t1 (t0+interval): failed ping
- t2: successful ping

At t2: 
- recreate subscription
- trigger store v3 query to get msg id of missed messages.
   -  either use last received message with timestamp < t0 as cursor (likely best practice)
   - or do a time based query: start: t0, end: t2

Reply:
I am curious to know how long does it take for websocket to detect timeout, because iirc websockets use TCP as underlying transport same as what is used in other waku implementations as well.
One possibility could be that websocket is having some sort of a very regular heartbeat which makes it detect quickly.

Reply:
> it is still experimental but on my local repros I observe the issue when subscription is dropped by a service node, that is happening somewhere after 10-15 mins.

The filter timeout has been set in nwaku to be 5 minutes, so it should get dropped after 5 minutes +/- some buffer time.

Reply:
@fryorcraken I would drop it out of this version of RFC
https://github.com/waku-org/specs/pull/18#discussion_r1663407256

this sounds like a more complex feature that should be useful but we don't have any precedent

what if I or someone from the team implements it in js-waku as part of improving Filter and after it we can add it to this RFC?